river/cluster/odac.py

```diff
@@ -0,0 +1,532 @@
+from __future__ import annotations
+
+import collections
+import itertools
+import math
+
+from river import base, stats
+
+
+class ODAC(base.Clusterer):
+    """ODAC: The Online Divisive-Agglomerative Clustering (ODAC) is an
+    algorithm whose the goal is to continuously maintain a hierarchical cluster's
+    structure from evolving time series data streams.
+
+    * Performs hierarchical clustering
+    * Continuously monitor the evolution of clusters' diameters
+    * Two Operators:
+        * Splitting: expand the structure more data, more detailed clusters
+        * Merge: contract the structure reacting to changes.
+    * Splitting and agglomerative criteria are supported by a confidence
+        level given by the Hoeffding bounds.
+
+    Main Algorithm
+
+    * ForEver
+        * Read Next Example
+        * For all the active clusters
+            * Update the sufficient statistics
+        * Time to Time
+            * Verify Merge Clusters
+            * Verify Expand Cluster
+
+    Feeding ODAC
+
+    * Each example is processed once.
+    * Only sufficient statistics at leaves are updated.
+    * Sufficient Statistics: a triangular matrix of the correlations between
+        variables in a leaf.
+
+    Similarity Distance
+
+    * Distance between time-series (a & b): rnomc(a,b) = sqrt((1-corr(a,b))/2) where
+        corr(a,b) is the Pearson Correlation coefficient.
+
+    The Merge Operator
+
+    The Splitting Criteria guarantees that cluster's diameters monotonically decrease.
+        * Assume Clusters: cj with descendants ck and cs.
+        * If diameter (ck) - diameter (cj) > ε OR diameter (cs) - diameter (cj ) > ε:
+            * Change in the correlation structure!
+            * Merge clusters ck and cs into cj.
+
+    Splitting Criteria
+
+    d1 = d(a,b) the farthest distance
+    d2 the second farthest distance
+
+    if d1 - d2 > εk or t > εk then
+        if (d1 - d0)|(d1 - d_avg) - (d_avg - d0) > εk then
+            * Split the cluster
+
+    Parameters
+    ----------
+    confidence_level
+        The confidence level that user wants to work.
+    n_min
+        Number of minimum observations to time to time verify Merge or Expand Clusters.
+    tau
+        The value of tau to use in the calculation of the Hoeffding bound.
+
+    Attributes
+    ----------
+    structure_changed : bool
+        This variable is true when the structure changed, produced by splitting or aggregation.
+
+    Examples
+    --------
+
+    In the following example we utilize one dataset described in paper [^1].
+
+    >>> from river.cluster import ODAC
+    >>> from river.stream import iter_csv
+
+    >>> model = ODAC(confidence_level=0.9, n_min=50)
+
+    >>> dataset = iter_csv("four_cluster_dataset.csv")
```

Comment:
Hi @Bezum30, we need to use another dataset for the tests. Maybe you can take a look at `datasets` or `datasets.synth`.